hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java

```diff
@@ -265,115 +259,41 @@ private Void addToBatchTransactionInfoWithTrace(String parentName,
-                          .setCurrentTerm(term).build());
-                  return null;
-                });
-
-            long startTime = Time.monotonicNow();
-            flushBatchWithTrace(lastTraceId.get(), readyBuffer.size(),
-                () -> {
-                  omMetadataManager.getStore().commitBatchOperation(
-                      batchOperation);
-                  return null;
-                });
-            ozoneManagerDoubleBufferMetrics.updateFlushTime(
-                Time.monotonicNow() - startTime);
-          }
-
-          // Complete futures first and then do other things. So, that
-          // handler threads will be released.
-          if (!isRatisEnabled) {
-            // Once all entries are flushed, we can complete their future.
-            readyFutureQueue.iterator().forEachRemaining((entry) -> {
-              entry.complete(null);
-            });
-
-            readyFutureQueue.clear();
-          }
-
-          int flushedTransactionsSize = readyBuffer.size();
-          flushedTransactionCount.addAndGet(flushedTransactionsSize);
-          flushIterations.incrementAndGet();
-
-          if (LOG.isDebugEnabled()) {
-            LOG.debug("Sync Iteration {} flushed transactions in this " +
-                    "iteration {}", flushIterations.get(),
-                flushedTransactionsSize);
-          }
-
-          // When non-HA do the sort step here, as the sorted list is not
-          // required for flush to DB. As in non-HA we want to complete
-          // futures as quick as possible after flush to DB, to release rpc
-          // handler threads.
-          if (!isRatisEnabled) {
-            flushedEpochs =
-                readyBuffer.stream().map(DoubleBufferEntry::getTrxLogIndex)
-                    .sorted().collect(Collectors.toList());
-          }
-
-
-          // Clean up committed transactions.
-
-          cleanupCache(cleanupEpochs);
-
-          readyBuffer.clear();
-
-          if (isRatisEnabled) {
-            releaseUnFlushedTransactions(flushedTransactionsSize);
-          }
-
-          // update the last updated index in OzoneManagerStateMachine.
-          ozoneManagerRatisSnapShot.updateLastAppliedIndex(
-              flushedEpochs);
+        // Wait till there is some transaction to flush.
+        canFlush();
+
+        setReadyBuffer();
+
+        // For snapshot, we want including all the keys that were committed
+        // before the snapshot `create` command was executed. To achieve
+        // the behaviour, we spilt the request buffer at snapshot create
+        // request and flush the buffer in batches split at snapshot create
+        // request.
+        // For example, if requestBuffer is [request1, request2,
+        // snapshotCreateRequest1, request3, snapshotCreateRequest2, request4].
+        //
+        // Split requestBuffer would be.
+        // bufferQueues = [[request1, request2], [snapshotRequest1, request3],
```

Comment:
Let us use snapshotCreate() as a barrier operation. For example, bufferQueues can be broken as  [[request1, request2], [snapshotRequest1] ,[request3], [snapshotRequest2], [request4,...]].

Reply:
As  discussed offline, it is better to use `snapshotCreate` as a barrier to eliminate the possibility of exposing race condition to RocksDB.

Updated as suggested in new revision.